crates/test-utils/src/revive.rs

```diff
@@ -0,0 +1,162 @@
+            .spawn()?;
+        // wait for rpc to be initialized
+        let mut attempts = 1;
+        loop {
+            sleep(RETRY_DELAY);
+            tracing::debug!(
+                "Connecting to contracts enabled node, attempt {}/{}",
+                attempts,
+                MAX_ATTEMPTS
+            );
+            match OnlineClient::<PolkadotConfig>::new().await {
+                Result::Ok(_) => return Ok(Self { node, tmp_dir }),
+                Err(err) => {
+                    if attempts < MAX_ATTEMPTS {
+                        attempts += 1;
+                        continue
+                    }
+                    let err = eyre::eyre!(
+                        "Failed to connect to node rpc after {} attempts: {}",
+                        attempts,
+                        err
+                    );
+                    tracing::error!("{}", err);
+                    node.kill()?;
+                    return Err(err);
+                }
+            }
+        }
+    }
+
+    fn kill(&mut self) {
+        tracing::debug!("Killing contracts node process {}", self.node.id());
+        if let Err(err) = self.node.kill() {
+            tracing::error!("Error killing contracts node process {}: {}", self.node.id(), err)
+        }
+    }
+}
+
+/// Spawn and manage an instance of an ethereum RPC proxy node.
+struct RpcProxyProcess(process::Child);
+
+impl Drop for RpcProxyProcess {
+    fn drop(&mut self) {
+        self.kill()
+    }
+}
+
+impl RpcProxyProcess {
+    async fn start() -> Result<Self> {
+        let mut rpc_proxy = process::Command::new(RPC_PROXY_BINARY)
+            .env("RUST_LOG", "error")
+            .arg("--dev")
+            .spawn()?;
+
+        let client = ClientBuilder::default().connect(RPC_URL).await?;
+
+        let mut attempts = 1;
+        loop {
+            sleep(RETRY_DELAY);
+            match client.request_noparams::<String>("eth_chainId").await {
+                Result::Ok(_) => {
+                    return Ok(Self(rpc_proxy));
+                }
+                Err(err) => {
+                    if attempts < MAX_ATTEMPTS {
+                        attempts += 1;
+                        continue
+                    }
+
+                    let err = eyre::eyre!(
+                        "Failed to connect to RPC proxy after {} attempts: {}",
+                        MAX_ATTEMPTS,
+                        err
+                    );
+                    tracing::error!("{}", err);
+                    rpc_proxy.kill()?;
+                    return Err(err);
+                }
+            }
+        }
+    }
+
+    fn kill(&mut self) {
+        tracing::debug!("Killing RPC proxy process {}", self.0.id());
+        if let Err(err) = self.0.kill() {
+            tracing::error!("Error killing RPC proxy process {}: {}", self.0.id(), err)
+        }
+    }
+}
+
+/// `PolkadotHubNode` combines a `substrate-node` with an Ethereum RPC proxy to enable
```

Comment:
should this then be feature-gated as not everyone will want/need to install subxt dep/substrate-node?

Reply:
The test will simply return if the dependencies are not met. I've added test example.